test/gtest/ucm/cuda_hooks.cc

```diff
@@ -63,66 +25,145 @@ class cuda_hooks : public ucs::test {
+
+    void check_event_present(
+            const mem_event_vec_t &events, const std::string &name, void *ptr,
+            size_t size,
+            ucs_memory_type_t mem_type = UCS_MEMORY_TYPE_UNKNOWN) const
+    {
+        for (const auto e : events) {
+            if (/* Start address match */
+                (ptr >= e.address) &&
+                /* End address match */
+                (UCS_PTR_BYTE_OFFSET(ptr, size) <=
+                 UCS_PTR_BYTE_OFFSET(e.address, e.size)) &&
+                /* Memory type match */
+                ((e.mem_type == mem_type) ||
+                 (e.mem_type == UCS_MEMORY_TYPE_UNKNOWN))) {
+                return;
+            }
+        }
+
+        FAIL() << "Could not file memory " << name << " event for " << ptr
+               << ".." << UCS_PTR_BYTE_OFFSET(ptr, size) << " type "
+               << ucs_memory_type_names[mem_type];
+    }
+
+    void mem_alloc_event(void *address, size_t size, ucs_memory_type_t mem_type)
+    {
+        m_alloc_events.push_back({address, size, mem_type});
+    }
+
+    void mem_free_event(void *address, size_t size)
+    {
+        m_free_events.push_back({address, size, UCS_MEMORY_TYPE_UNKNOWN});
+    }
+
+    static void cuda_vm_mapped_callback(ucm_event_type_t event_type,
+                                        ucm_event_t *event, void *arg)
+    {
+        auto self = reinterpret_cast<cuda_hooks*>(arg);
+        self->mem_alloc_event(event->vm_mapped.address, event->vm_mapped.size,
+                              UCS_MEMORY_TYPE_HOST);
+    }
+
+    static void cuda_vm_unmapped_callback(ucm_event_type_t event_type,
+                                          ucm_event_t *event, void *arg)
+    {
+        auto self = reinterpret_cast<cuda_hooks*>(arg);
+        self->mem_free_event(event->vm_unmapped.address,
+                             event->vm_unmapped.size);
+    }
+
+    static void cuda_mem_alloc_callback(ucm_event_type_t event_type,
+                                        ucm_event_t *event, void *arg)
+    {
+        auto self = reinterpret_cast<cuda_hooks*>(arg);
+        self->mem_alloc_event(event->mem_type.address, event->mem_type.size,
+                              event->mem_type.mem_type);
+    }
+
+    static void cuda_mem_free_callback(ucm_event_type_t event_type,
+                                       ucm_event_t *event, void *arg)
     {
-        ASSERT_EQ(ptr, alloc_event.address);
-        ASSERT_GE(alloc_event.size, size);
-        EXPECT_TRUE((alloc_event.mem_type == expect_mem_type) ||
-                    (alloc_event.mem_type == UCS_MEMORY_TYPE_UNKNOWN));
+        auto self = reinterpret_cast<cuda_hooks*>(arg);
+        self->mem_free_event(event->mem_type.address, event->mem_type.size);
     }
 
-    CUdevice   device;
-    CUcontext  context;
+    CUdevice        m_device = CU_DEVICE_INVALID;
+    CUcontext       m_context = NULL;
+    mem_event_vec_t m_alloc_events;
+    mem_event_vec_t m_free_events;
```

Comment:
Minor:

```suggestion
    CUdevice        m_device{CU_DEVICE_INVALID};
    CUcontext       m_context{NULL};
    mem_event_vec_t m_alloc_events{1000};
    mem_event_vec_t m_free_events{1000};
```

Instead of calling `reserve()` for vectors during the initialization, we can initialize all the members in place of definition (looks more consistent from my point of view).

Reply:
@ivankochin but if we initialize the vectors to 1000 it will be the size, not the capacity..

Reply:
Yes, you are right, completely forgot about that, sorry for confusion!